File: 0213-house-robber-ii/0213-house-robber-ii.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function (nums) {

    const solve = (arr) => {
        const n = arr.length;
        let prev = arr[0];
        let prev2 = 0;

        for (let i = 1; i < n; i++) {
            let pick = arr[i];

            if (i > 1)
                pick += prev2;

            let notPick = 0 + prev;

            let cur_i = Math.max(pick, notPick);
            prev2 = prev;
            prev = cur_i;
        }

        return prev;
    };

    const n = nums.length;
    let temp1 = [];
    let temp2 = [];

    if(n === 1) return nums[0];

    for (let i = 0; i < nums.length; i++) {
        if (i != n - 1)
            temp1.push(nums[i]);

        if (i !== 0)
            temp2.push(nums[i]);
    }

    return Math.max(solve(temp1), solve(temp2));
};

File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:

        n = len(nums)
        if n == 0:
            return 0

        if n == 1:
            return nums[0]

        temp1 = []
        temp2 = []

        for i in range(n):
            if i != n - 1:
                temp1.append(nums[i])

            if i != 0:
                temp2.append(nums[i])

        ans1 = self._helper(temp1)
        ans2 = self._helper(temp2)

        return max(ans1, ans2)

    def _helper(self, nums: List[int]):

        n = len(nums)

        if n == 0:
            return 0
        if n == 1:
            return nums[0]

        prev2 = nums[0]
        prev = max(nums[0], nums[1])

        for i in range(2, n):
            rob = nums[i] + prev2
            no_rob = prev
            curi = max(rob, no_rob)

            prev2 = prev
            prev = curi

        return prev
